File: scripts/transport-v4-rekey-credit.mjs
import assert from "node:assert/strict";
import { types } from "node:util";
import { VectorError } from "./transport-v4-codec.mjs";
import { verifyCryptoUsageRegistry } from "./transport-v4-crypto-usage.mjs";
import { timeArithmeticReference } from "./transport-v4-time.mjs";

const requireThat = (condition, code) => { if (!condition) throw new VectorError(code); };
const rateFields = ["rate_numerator", "rate_denominator", "quantization_ms"];
const creditFields = ["burst_rounds", "refill_period_ms", "base_credit", "delta_ms", ...rateFields];
const operations = {
  service: ["burst_rounds", "refill_period_ms", "request_start_budget_ms", "issued_at_ms", "session_not_after_ms", ...rateFields],
  initial_credit: ["burst_rounds", "refill_period_ms"],
  client_credit: creditFields,
  server_credit: creditFields,
};

export function verifyRekeyCreditRegistry(schema) {
  const registry = schema.rekey_credit_registry;
  assert.equal(registry.status, "reference_arithmetic_only");
  assert.equal(registry.source, "3.8");
  assert.equal(registry.quantity_max, schema.time_arithmetic_registry.quantity_max);
  assert.equal(registry.intermediate_max, schema.time_arithmetic_registry.intermediate_max);
  assert.deepEqual(registry.operations, operations);
  assert.equal(registry.additional_authorization_tolerance_ms, "0");
  assert.equal(registry.service_error, "configuration_capacity");
  assert.equal(registry.error_allowance, "ceil(2*eta/(1-rho))+1");
  assert.equal(registry.service_rounds, "floor((B*R+B*((1+rho)/(1-rho))*T)/(R-B*e))");
  assert.deepEqual(registry.elapsed_bound, { client_credit: "elapsed_lower_ms", server_credit: "elapsed_upper_ms" });
  assert.match(registry.qualification, /no live owner/u);
  verifyCryptoUsageRegistry(schema);
  const fields = schema.frame_maps.RekeyEnvelope.fields;
  assert.deepEqual(Object.values(fields).map(field => [field.name, field.type, field.min]),
    [["burst_rounds", "uint16", 1], ["refill_period_ms", "uint32", 1], ["request_start_budget_ms", "uint32", 1]]);
}

// Pure decimal-string projections; neither a credit mutation nor an INIT permit.
// Runtime owns the original authenticated Artifact/TimeProfile, retained base,
// causal ACK anchor, once-only INIT deduction and real service/cleanup reserves.
// A pending round cannot accrue credit; ACK preserves post_charge unchanged.
export function rekeyCreditReference(schema, profile, operation, input) {
  verifyRekeyCreditRegistry(schema);
  const registry = schema.rekey_credit_registry;
  requireThat(typeof profile === "string" && Object.hasOwn(schema.crypto_usage_registry.profiles, profile), "rekey_credit_profile");
  requireThat(typeof operation === "string" && Object.hasOwn(registry.operations, operation), "rekey_credit_operation");
  requireThat(input !== null && typeof input === "object" && !types.isProxy(input) &&
    [Object.prototype, null].includes(Object.getPrototypeOf(input)), "rekey_credit_input");
  const names = registry.operations[operation], keys = Reflect.ownKeys(input);
  requireThat(keys.length === names.length && keys.every(key => names.includes(key)), "rekey_credit_fields");
  const maximum = BigInt(registry.quantity_max), wideMaximum = BigInt(registry.intermediate_max);
  const values = Object.create(null);
  for (const name of names) {
    const descriptor = Object.getOwnPropertyDescriptor(input, name);
    requireThat(descriptor && Object.hasOwn(descriptor, "value") && descriptor.enumerable, "rekey_credit_data");
    const value = descriptor.value;
    requireThat(typeof value === "string" && value.length <= registry.quantity_max.length && /^(?:0|[1-9][0-9]*)$/u.test(value), "rekey_credit_integer");
    values[name] = BigInt(value);
    requireThat(values[name] <= maximum, "rekey_credit_integer");
  }
  const capacity = condition => requireThat(condition, registry.service_error);
  const wide = value => { capacity(value >= 0n && value <= wideMaximum); return value; };
  const ceil = (a, b) => a / b + (a % b === 0n ? 0n : 1n);
  for (const field of Object.values(schema.frame_maps.RekeyEnvelope.fields)) {
    if (!Object.hasOwn(values, field.name)) continue;
    const bits = Number(field.type.slice(4));
    capacity(values[field.name] >= BigInt(field.min) && values[field.name] < (1n << BigInt(bits)));
  }
  const B = values.burst_rounds, R = values.refill_period_ms, C = wide(B * R);
  const maxEpochs = BigInt(schema.crypto_usage_registry.profiles[profile].max_epochs);
  capacity(B < maxEpochs);
  let result;
  if (operation === "service") {
    const { rate_numerator: n, rate_denominator: d, quantization_ms: q } = values;
    capacity(d > 0n && n < d && values.issued_at_ms < values.session_not_after_ms);
    const minus = d - n, plus = wide(d + n), T = values.session_not_after_ms - values.issued_at_ms;
    // R>B*e bounds e before computing 2*q*d, avoiding an overflowing
    // intermediate for inputs that cannot possibly fit the service envelope.
    const maxError = (R - 1n) / B;
    capacity(maxError >= 1n);
    capacity(q <= wide((maxError - 1n) * minus) / wide(2n * d));
    const e = ceil(wide(wide(2n * q) * d), minus) + 1n;
    const denominatorMs = R - B * e;
    capacity(denominatorMs > 0n);
    const denominator = wide(denominatorMs * minus), initial = wide(C * minus), coefficient = wide(B * plus);
    // floor(numerator/denominator)<maxEpochs iff numerator<maxEpochs*denominator.
    // Bound T first. Accepted products then fit uint128 even for uint64 rates.
    const exclusiveLimit = wide(maxEpochs * denominator);
    capacity(initial < exclusiveLimit);
    capacity(T <= (exclusiveLimit - 1n - initial) / coefficient);
    const numerator = wide(initial + wide(coefficient * T));
    const rounds = numerator / denominator;
    capacity(rounds > 0n && rounds < maxEpochs);
    result = { service_ms: T, error_allowance_ms: e, denominator_ms: denominatorMs,
      capacity_credit: C, max_rounds: rounds, required_epochs: rounds + 1n };
  } else {
    let available = C;
    if (operation !== "initial_credit") {
      capacity(values.base_credit <= C);
      const elapsed = timeArithmeticReference(schema, "elapsed", Object.fromEntries(
        ["delta_ms", ...rateFields].map(name => [name, values[name].toString()])));
      const u = BigInt(elapsed[registry.elapsed_bound[operation]]);
      // Saturate before multiplication; never multiply a huge measured idle.
      if (u < R) {
        const refill = wide(B * u);
        available = refill >= C - values.base_credit ? C : values.base_credit + refill;
      }
    }
    result = { capacity_credit: C, available_credit: available, post_charge_credit: available >= R ? available - R : null };
  }
  return Object.freeze(Object.fromEntries(Object.entries(result).map(([key, value]) => [key, value === null ? null : value.toString()])));
}

export function buildRekeyCreditCorpus(schema) {
  verifyRekeyCreditRegistry(schema);
  const ids = new Set(), vectors = [];
  for (const profile of Object.keys(schema.crypto_usage_registry.profiles)) {
    for (const spec of schema.rekey_credit_vector_plan) {
      assert.match(spec.id, /^[a-z][a-z0-9_]*$/u);
      const id = `rekey_credit_${schema.crypto_profiles[profile].dh_algorithm}_${spec.id}`;
      assert.ok(!ids.has(id)); ids.add(id);
      const input = structuredClone(spec.input);
      if (spec.expected_error) {
        assert.equal(spec.expected, undefined);
        assert.throws(() => rekeyCreditReference(schema, profile, spec.operation, input), error => error instanceof VectorError && error.code === spec.expected_error, id);
        vectors.push({ id, profile, operation: spec.operation, input, expected_error: spec.expected_error });
      } else {
        const expected = rekeyCreditReference(schema, profile, spec.operation, input);
        assert.deepEqual(expected, spec.expected, id);
        vectors.push({ id, profile, operation: spec.operation, input, expected });
      }
    }
  }
  return { status: "draft", schema_revision: schema.schema_revision, design_sha256: schema.design_sha256,
    coverage: "rekey_full_session_service_and_credit_arithmetic_only", qualification: schema.rekey_credit_registry.qualification, vectors };
}

File: flowersec-rust/tests/transport_v4_rekey_credit.rs
//! Independent test arithmetic; no live INIT/ACK owner or service reservation.
#[allow(dead_code)]
#[path = "../src/protocol_v4_registry_generated.rs"]
mod registry;
#[allow(dead_code)]
#[path = "support/v4_time.rs"]
mod time;
use serde_json::{Map, Value, json};
use std::collections::BTreeMap;

fn compute(profile: &str, operation: &str, input: &Value) -> Result<Value, &'static str> {
    let spec: Value = serde_json::from_str(registry::REKEY_CREDIT_REGISTRY_JSON).unwrap();
    let usage: Value = serde_json::from_str(registry::CRYPTO_USAGE_REGISTRY_JSON).unwrap();
    let time_spec: Value = serde_json::from_str(registry::TIME_ARITHMETIC_REGISTRY_JSON).unwrap();
    let profile = usage["profiles"]
        .get(profile)
        .ok_or("rekey_credit_profile")?;
    let fields = spec["operations"][operation]
        .as_array()
        .ok_or("rekey_credit_operation")?;
    let object = input.as_object().ok_or("rekey_credit_input")?;
    if object.len() != fields.len()
        || object
            .keys()
            .any(|k| !fields.iter().any(|f| f.as_str() == Some(k)))
    {
        return Err("rekey_credit_fields");
    }
    let mut values = BTreeMap::new();
    for field in fields {
        let key = field.as_str().unwrap();
        let s = object[key].as_str().ok_or("rekey_credit_integer")?;
        if s.is_empty()
            || s.len() > spec["quantity_max"].as_str().unwrap().len()
            || s.len() > 1 && s.starts_with('0')
            || !s.bytes().all(|b| b.is_ascii_digit())
        {
            return Err("rekey_credit_integer");
        }
        values.insert(
            key,
            s.parse::<u64>().map_err(|_| "rekey_credit_integer")? as u128,
        );
    }
    let v = |key: &str| values[key];
    let cap = "configuration_capacity";
    let wide = spec["intermediate_max"]
        .as_str()
        .unwrap()
        .parse::<u128>()
        .unwrap();
    let mul = |a: u128, b: u128| a.checked_mul(b).filter(|n| *n <= wide).ok_or(cap);
    for field in spec["envelope_fields"].as_object().unwrap().values() {
        let key = field["name"].as_str().unwrap();
        if let Some(n) = values.get(key) {
            let bits = field["type"].as_str().unwrap()[4..].parse::<u32>().unwrap();
            if *n < field["min"].as_u64().unwrap() as u128 || *n >= (1u128 << bits) {
                return Err(cap);
            }
        }
    }
    let (b, r) = (v("burst_rounds"), v("refill_period_ms"));
    let capacity = mul(b, r)?;
    let epochs = profile["max_epochs"]
        .as_str()
        .unwrap()
        .parse::<u128>()
        .unwrap();
    if b >= epochs {
        return Err(cap);
    }
    if operation == "service" {
        let (n, d, q) = (
            v("rate_numerator"),
            v("rate_denominator"),
            v("quantization_ms"),
        );
        if d == 0 || n >= d || v("issued_at_ms") >= v("session_not_after_ms") {
            return Err(cap);
        }
        let (dm, dp, duration) = (d - n, d + n, v("session_not_after_ms") - v("issued_at_ms"));
        let max_error = (r - 1) / b;
        if max_error == 0 || q > mul(max_error - 1, dm)? / mul(2, d)? {
            return Err(cap);
        }
        let e = mul(mul(2, q)?, d)?.div_ceil(dm) + 1;
        let period = r.checked_sub(mul(b, e)?).filter(|x| *x > 0).ok_or(cap)?;
        let (denominator, initial, rate) = (mul(period, dm)?, mul(capacity, dm)?, mul(b, dp)?);
        let limit = mul(epochs, denominator)?;
        if initial >= limit || duration > (limit - 1 - initial) / rate {
            return Err(cap);
        }
        let numerator = initial
            .checked_add(mul(rate, duration)?)
            .filter(|n| *n <= wide)
            .ok_or(cap)?;
        let rounds = numerator / denominator;
        if rounds == 0 || rounds >= epochs {
            return Err(cap);
        }
        return Ok(
            json!({"service_ms":duration.to_string(),"error_allowance_ms":e.to_string(),
            "denominator_ms":period.to_string(),"capacity_credit":capacity.to_string(),
            "max_rounds":rounds.to_string(),"required_epochs":(rounds+1).to_string()}),
        );
    }
    let mut available = capacity;
    if operation != "initial_credit" {
        let base = v("base_credit");
        if base > capacity {
            return Err(cap);
        }
        let input: Map<String, Value> = [
            "delta_ms",
            "rate_numerator",
            "rate_denominator",
            "quantization_ms",
        ]
        .into_iter()
        .map(|k| (k.to_owned(), json!(v(k).to_string())))
        .collect();
        let elapsed = time::compute(&time_spec, "elapsed", &Value::Object(input))?;
        let bound = spec["elapsed_bound"][operation].as_str().unwrap();
        let u = elapsed[bound].as_str().unwrap().parse::<u128>().unwrap();
        if u < r {
            let refill = mul(b, u)?;
            if refill < capacity - base {
                available = base + refill;
            }
        }
    }
    let post = available.checked_sub(r).map(|n| n.to_string());
    Ok(
        json!({"capacity_credit":capacity.to_string(),"available_credit":available.to_string(),"post_charge_credit":post}),
    )
}

#[test]
fn every_shared_rekey_credit_vector() {
    let corpus: Value = serde_json::from_str(include_str!(
        "../../testdata/transport_v4/rekey_credit.json"
    ))
    .unwrap();
    let vectors = corpus["vectors"].as_array().unwrap();
    assert_eq!(vectors.len(), 48);
    for v in vectors {
        let before = v["input"].clone();
        let result = compute(
            v["profile"].as_str().unwrap(),
            v["operation"].as_str().unwrap(),
            &before,
        );
        if let Some(error) = v["expected_error"].as_str() {
            assert_eq!(result, Err(error), "{}", v["id"]);
        } else {
            assert_eq!(result, Ok(v["expected"].clone()), "{}", v["id"]);
        }
        assert_eq!(before, v["input"]);
    }
}

#[test]
fn floor_boundaries_partial_credit_and_strict_inputs() {
    for duration in 1u64..=120 {
        let result = compute(registry::DH_PROFILE_X25519,"service",&json!({"burst_rounds":"2","refill_period_ms":"30",
            "request_start_budget_ms":"5","issued_at_ms":"0","session_not_after_ms":duration.to_string(),
            "rate_numerator":"1","rate_denominator":"2","quantization_ms":"2"})).unwrap();
        let n = result["max_rounds"]
            .as_str()
            .unwrap()
            .parse::<u64>()
            .unwrap();
        assert!(n * 12 <= 60 + 6 * duration && (n + 1) * 12 > 60 + 6 * duration);
    }
    for bad in [
        json!(1),
        json!(true),
        Value::Null,
        json!(""),
        json!("01"),
        json!("-1"),
        json!("+1"),
        json!("1.0"),
        json!("1\n"),
        json!("١"),
        json!("18446744073709551616"),
    ] {
        assert_eq!(
            compute(
                registry::DH_PROFILE_X25519,
                "initial_credit",
                &json!({"burst_rounds":bad,"refill_period_ms":"30"})
            ),
            Err("rekey_credit_integer")
        );
    }
    let mut input = json!({"burst_rounds":"2","refill_period_ms":"30","base_credit":"10","delta_ms":"9",
        "rate_numerator":"0","rate_denominator":"1","quantization_ms":"0"});
    for _ in 0..50 {
        let peek = compute(registry::DH_PROFILE_X25519, "client_credit", &input).unwrap();
        assert_eq!(peek["available_credit"], "28");
        assert!(peek["post_charge_credit"].is_null());
    }
    input["delta_ms"] = json!("10");
    assert_eq!(
        compute(registry::DH_PROFILE_X25519, "client_credit", &input).unwrap()["post_charge_credit"],
        "0"
    );
}
